akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -64,14 +95,38 @@ object MergeHub {
    */
   def source[T]: Source[T, Sink[T, NotUsed]] = source(perProducerBufferSize = 16)
 
+  /**
+   * Creates a [[Source]] that emits elements merged from a dynamic set of producers. After the [[Source]] returned
+   * by this method is materialized, it returns a [[Sink]] as a materialized value. This [[Sink]] can be materialized
+   * arbitrary many times and each of the materializations will feed the elements into the original [[Source]].
+   *
+   * Every new materialization of the [[Source]] results in a new, independent hub, which materializes to its own
+   * [[Sink]] for feeding that materialization.
+   *
+   * Completed or failed [[Sink]]s are simply removed. Once the [[Source]] is cancelled, the Hub is considered closed
+   * and any new producers using the [[Sink]] will be cancelled.
+   *
+   * The materialized [[DrainingControl]] can be used to drain the Hub: any new produces using the [[Sink]] will be cancelled
+   * and the Hub will be closed completing the [[Source]] as soon as all currently connected producers complete.
+   */
+  def sourceWithDraining[T](): Source[T, (Sink[T, NotUsed], DrainingControl)] =
+    sourceWithDraining(perProducerBufferSize = 16)
+
   final class ProducerFailed(msg: String, cause: Throwable) extends RuntimeException(msg, cause)
 }
 
 /**
  * INTERNAL API
  */
-private[akka] class MergeHub[T](perProducerBufferSize: Int)
-    extends GraphStageWithMaterializedValue[SourceShape[T], Sink[T, NotUsed]] {
+@InternalApi
+private[akka] final class MergeHubDrainingControlImpl(drainAction: () => Unit) extends MergeHub.DrainingControl {
```

Comment:
if we keep the name `DrainingControl`, the impl can be called `DrainingControlImpl`. Like in the Java API.

Reply:
A common pattern we have is to have one impl class in an internal package that implements both Scala and Java interface. 

For example

https://github.com/akka/akka/blob/c5655a9ce605e3eb9982e8dd4d1e5a2c7d327e5e/akka-actor-typed/src/main/scala/akka/actor/typed/internal/StashBufferImpl.scala#L39-L45

But it seems that akka-stream is not following that pattern and the implementations are in `akka.stream.impl`.

Maybe the best here is to add the following in the `akka.stream.impl` package. 


```scala
private[akka] final class MergeHubDrainingControlImpl(drainAction: () => Unit) extends scaladsl.MergeHub.DrainingControl with javadsl.MergeHub.DrainingControl {
  override def drainAndComplete(): Unit = {
    drainAction()
  }
}
```

Note that in that case, it's better to preserve the name `MergeHubDrainingControlImpl` because it won't be under `MergeHub` anymore. 

We end up with one impl class that is good for both APIs and we don't need a delegate anymore.